File: app/src/main/java/com/vikho305/isaho220/outstanding/data/repositories/UserRepository.java
package com.vikho305.isaho220.outstanding.data.repositories;

import android.content.Context;

import com.vikho305.isaho220.outstanding.data.AuthInfo;
import com.vikho305.isaho220.outstanding.data.Post;
import com.vikho305.isaho220.outstanding.data.User;
import com.vikho305.isaho220.outstanding.data.api.UserApi;

import java.util.List;

import io.reactivex.Single;

public class UserRepository {

    private UserApi userApi;
    private PreferenceRepository preferences;

    public UserRepository(Context context) {
        userApi = new UserApi(context);
        preferences = new PreferenceRepository(context);
    }

    public Single<String> register(String username, String email, String password) {
        return userApi.register(username, email, password);
    }

    public Single<AuthInfo> login(String username, String password) {
        return userApi.login(username, password);
    }

    public Single<String> logout() {
        return userApi.logout();
    }

    // Getter methods

    public Single<User> getUser(String userId) {
        return userApi.getUser(userId);
    }

    public Single<User> getUserSelf() {
        return userApi.getUser(preferences.getAuthUserId());
    }

    // TODO: add more self methods

    public Single<List<User>> getFollowers(String userId, int page) {
        return userApi.getFollowers(userId, page);
    }

    public Single<List<User>> getFollowings(String userId, int page) {
        return userApi.getFollowings(userId, page);
    }

    public Single<List<User>> getPendingFollowers(String userId, int page) {
        return userApi.getPendingFollowers(userId, page);
    }

    public Single<List<User>> getPendingFollowings(String userId, int page) {
        return userApi.getPendingFollowings(userId, page);
    }

    public Single<List<User>> getBlockings(String userId, int page) {
        return userApi.getBlockings(userId, page);
    }

    public Single<List<Post>> getPosts(String userId, int page) {
        return userApi.getPosts(userId, page);
    }

    // Setter methods

    public Single<String> setAccountDetails(String username, String email) {
        return userApi.setAccountDetails(username, email);
    }

    public Single<String> setAccountDetails(String username, String email, String password) {
        return userApi.setAccountDetails(username, email, password);
    }

    public Single<String> setProfileDetails(String picture, String description,
                                            double primaryH, double primaryS, double primaryL,
                                            double secondaryH, double secondaryS, double secondaryL) {
        return userApi.setProfileDetails(picture, description,
                primaryH, primaryS, primaryL, secondaryH, secondaryS, secondaryL);
    }

    public Single<String> setCoordinates(double latitude, double longitude) {
        return userApi.setCoordinates(latitude, longitude);
    }

    // User interaction methods

    public Single<String> followUser(String userId) {
        return userApi.followUser(userId);
    }

    public Single<String> unfollowUser(String userId) {
        return userApi.unfollowUser(userId);
    }

    public Single<String> blockUser(String userId) {
        return userApi.blockUser(userId);
    }

    public Single<String> unblockUser(String userId) {
        return userApi.unblockUser(userId);
    }

    public Single<String> acceptFollower(String followerId) {
        return userApi.acceptFollower(followerId);
    }

    public Single<String> rejectFollower(String followerId) {
        return userApi.rejectFollower(followerId);
    }
}

File: app/src/main/java/com/vikho305/isaho220/outstanding/data/repositories/UserRepository.kt
package com.vikho305.isaho220.outstanding.data.repositories

import android.content.Context
import com.vikho305.isaho220.outstanding.data.api.UserApi
import com.vikho305.isaho220.outstanding.data.AuthInfo
import com.vikho305.isaho220.outstanding.data.Post
import com.vikho305.isaho220.outstanding.data.User
import io.reactivex.Single

class UserRepository(context: Context) {
    private val userApi: UserApi = UserApi(context)
    private val preferences: PreferenceRepository = PreferenceRepository(context)

    fun register(username: String?, email: String?, password: String?): Single<String> {
        return userApi.register(username, email, password)
    }

    fun login(username: String?, password: String?): Single<AuthInfo> {
        return userApi.login(username, password)
    }

    fun logout(): Single<String> {
        return userApi.logout()
    }

    /////////////////
    // Getter methods

    fun getUser(userId: String): Single<User> {
        return userApi.getUser(userId)
    }

    fun getUserSelf(): Single<User> {
        return userApi.getUser(preferences.authUserId!!)
    }

    fun getFollowers(userId: String, page: Int): Single<List<User>> {
        return userApi.getFollowers(userId, page)
    }

    fun getFollowings(userId: String, page: Int): Single<List<User>> {
        return userApi.getFollowings(userId, page)
    }

    fun getPendingFollowers(userId: String, page: Int): Single<List<User>> {
        return userApi.getPendingFollowers(userId, page)
    }

    fun getPendingFollowings(userId: String, page: Int): Single<List<User>> {
        return userApi.getPendingFollowings(userId, page)
    }

    fun getBlockings(userId: String, page: Int): Single<List<User>> {
        return userApi.getBlockings(userId, page)
    }

    fun getPosts(userId: String, page: Int): Single<List<Post>> {
        return userApi.getPosts(userId, page)
    }

    /////////////////
    // Setter methods

    fun setAccountDetails(username: String?, email: String?): Single<String> {
        return userApi.setAccountDetails(username, email)
    }

    fun setAccountDetails(username: String?, email: String?, password: String?): Single<String> {
        return userApi.setAccountDetails(username, email, password)
    }

    fun setProfileDetails(picture: String?, description: String?,
                          primaryH: Double, primaryS: Double, primaryL: Double,
                          secondaryH: Double, secondaryS: Double, secondaryL: Double): Single<String> {
        return userApi.setProfileDetails(picture, description,
                primaryH, primaryS, primaryL, secondaryH, secondaryS, secondaryL)
    }

    fun setCoordinates(latitude: Double, longitude: Double): Single<String> {
        return userApi.setCoordinates(latitude, longitude)
    }

    ///////////////////////////
    // User interaction methods

    fun followUser(userId: String): Single<String> {
        return userApi.followUser(userId)
    }

    fun unfollowUser(userId: String): Single<String> {
        return userApi.unfollowUser(userId)
    }

    fun blockUser(userId: String): Single<String> {
        return userApi.blockUser(userId)
    }

    fun unblockUser(userId: String): Single<String> {
        return userApi.unblockUser(userId)
    }

    fun acceptFollower(followerId: String): Single<String> {
        return userApi.acceptFollower(followerId)
    }

    fun rejectFollower(followerId: String): Single<String> {
        return userApi.rejectFollower(followerId)
    }
}